### PR TITLE
Add eu option, DEV-219

### LINKED ISSUE #22: European CDN solution
Guten Morgen,

wäre es möglich in den Einstellungen auch eine Einstellung für die European CDN solution hinzuzufügen.

https://support.cookiebot.com/hc/en-us/articles/4530208762396-Cookiebot-CMP-European-CDN-solution

Mfg

### PR DESCRIPTION
**Pull request by iranimij**

### CODE REVIEW

File: Test/Integration/Model/ScriptGeneratorTest.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    private $config;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+        $this->config = TestFrameworkObjectManager::getInstance()->create(Config::class);
+    }
+
+    public function testEuCdnEnabledReturnsFalseByDefault(): void
```
Comment (sprankhub) [inline Test/Integration/Model/ScriptGeneratorTest.php]: Please move this to a `ConfigTest` class.

File: Test/Integration/Model/ScriptGeneratorTest.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        self::assertFalse($this->config->isEuCdn());
+    }
+
+    /**
+     * @magentoConfigFixture default/web/cookiebot/enabled_eu_cdn 0
+     */
+    public function testIsEuCdndReturnsFalse(): void
```
Comment (sprankhub) [inline Test/Integration/Model/ScriptGeneratorTest.php]: Good tests :+1: Feedback applies to this and the next test:

1. Please improve the naming. This does not test if `IsEuCdndReturnsFalse`. I think something like `testEuCdnIsUsedWhenDisabled` would be much more descriptive.
2. I think `AddScriptTest` should be extended by the EU option (each test can be run with EU enabled / disabled), so that it also tests if the right script is actually inserted on the page.
3. I don't think this test would currently pass as `%s` is replaced.

File: etc/adminhtml/system.xml [hunk 3]
```diff
@@ -21,6 +21,10 @@
                         <field id="enabled">1</field>
                     </depends>
                 </field>
+                <field id="is_eu_cdn" translate="label" type="select" sortOrder="40" showInDefault="1" showInWebsite="1" showInStore="1">
```
Comment (sprankhub) [inline etc/adminhtml/system.xml]: Please let's rename this here and everywhere else to `use_eu_cdn`. Make sure to also use the right ID in the integration tests ;-)

File: Model/Config.php [hunk 4]
```diff
@@ -37,4 +38,9 @@ public function getDataCulture(): string
     {
         return (string)$this->scopeConfig->getValue(self::XML_PATH_DATA_CULTURE, ScopeInterface::SCOPE_STORE);
     }
+
+    public function isEuCdn(): bool
+    {
+        return (bool)$this->scopeConfig->getValue(self::XML_PATH_IS_EU_CDN, ScopeInterface::SCOPE_STORE);
```
Comment (sprankhub) [inline Model/Config.php]: Please use `isSetFlag` and remove the type cast.

**Review by sprankhub — CHANGES_REQUESTED** [review]

File: Test/Integration/Model/ScriptGeneratorTest.php (hunk 2 above)
Comment (iranimij) [reply Test/Integration/Model/ScriptGeneratorTest.php]: It's fixed, my tests are completely ok, but old tests return some errors, and we can debug them If you think in this task is a good place for fixing them please let me know

File: Model/Config.php [hunk 5]
```diff
@@ -37,4 +38,9 @@ public function getDataCulture(): string
     {
         return (string)$this->scopeConfig->getValue(self::XML_PATH_DATA_CULTURE, ScopeInterface::SCOPE_STORE);
     }
+
+    public function isEuCdn(): bool
```
Comment (sprankhub) [inline Model/Config.php]: Please rename to `useEuCdn` as well.

File: Test/Integration/Model/ConfigTest.php [hunk 6]
```diff
@@ -20,6 +20,11 @@ protected function setUp(): void
         $this->config = TestFrameworkObjectManager::getInstance()->create(Config::class);
     }
 
+    public function testEuCdnEnabledReturnsFalseByDefault(): void
```
Comment (sprankhub) [inline Test/Integration/Model/ConfigTest.php]: Please move this to the bottom to match the order in the other files. Also, please also test if the right value is returned if a value is set.

File: Test/Integration/AddScriptTest.php [hunk 7]
```diff
@@ -23,43 +23,55 @@ protected function setUp()
     /**
      * @magentoConfigFixture default/web/cookiebot/enabled 1
      * @magentoConfigFixture default/web/cookiebot/id 123-456-789
+     * @magentoConfigFixture current_store web/cookiebot/use_eu_cdn 1
      */
     public function testScriptAddedOnHomepage(): void
     {
         $this->dispatch('/');
         self::assertContains($this->script, $this->getResponse()->getBody());
+        self::assertEquals('<script id="Cookiebot" data-cfasync="false" src="https://consent.cookiebot.eu/uc.js" data-cbid=""  type="text/javascript" async></script>',
```
Comment (sprankhub) [inline Test/Integration/AddScriptTest.php]: I don't think we need to test this here, since we have such tests for the `ScriptGenerator`.

File: Test/Integration/Model/ScriptGeneratorTest.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        $this->config = TestFrameworkObjectManager::getInstance()->create(Config::class);
+    }
+
+    /**
+     * @magentoConfigFixture current_store web/cookiebot/use_eu_cdn 0
+     */
+    public function testWhenUseEuCdnIsDisabled(): void
+    {
+        $generatedScript = ObjectManager::getInstance()->create(ScriptGenerator::class)->generate();
+        self::assertEquals('<script id="Cookiebot" data-cfasync="false" src="https://consent.cookiebot.com/uc.js" data-cbid=""  type="text/javascript" async></script>',
```
Comment (sprankhub) [inline Test/Integration/Model/ScriptGeneratorTest.php]: I don't think this can pass as it includes a duplicate whitespace after the data-cbid attribute, which should not be in the generated script?
Anyway, since only the domain changes, I think it would be enough to check if the right domain is included in the script?

**Review by sprankhub — CHANGES_REQUESTED** [review]

File: Test/Integration/Model/ScriptGeneratorTest.php (hunk 2 above)
Comment (sprankhub) [reply Test/Integration/Model/ScriptGeneratorTest.php]: All tests should pass :)

File: Test/Integration/Model/ScriptGeneratorTest.php (hunk 8 above)
Comment (iranimij) [reply Test/Integration/Model/ScriptGeneratorTest.php]: No, there should be exactly two space to be passed, you mean I should first get the url by regex and then check the URL? Don't you think in this way the performance decreased and we should first get the URL?

File: Test/Integration/AddScriptTest.php (hunk 7 above)
Comment (iranimij) [reply Test/Integration/AddScriptTest.php]: Actually I remember in your previous comments you also asked me to add these tests here as well

Comment (iranimij) [reply Test/Integration/AddScriptTest.php]: https://github.com/customgento/module-cookiebot-m2/pull/23/files/461bda69f89c9927171ea1daa8ac01f25e141abb#r1404128423 second part

File: Test/Integration/Model/ConfigTest.php [hunk 9]
```diff
@@ -47,10 +47,15 @@ public function testGetIdReturnsEmptyStringByDefault(): void
...
     /**
-     * @magentoConfigFixture default/web/cookiebot/id 123-456-789
+     * @magentoConfigFixture current_store web/cookiebot/id 123-456-789
      */
     public function testGetIdReturnsId(): void
     {
         self::assertEquals('123-456-789', $this->config->getId());
     }
+
+    public function testEuCdnEnabledReturnsFalseByDefault(): void
```
Comment (sprankhub) [inline Test/Integration/Model/ConfigTest.php:57]: Similar to the other methods, we should also test if this method returns true if the config is true and returns false if the config is false.

File: Test/Integration/AddScriptTest.php (hunk 7 above)
Comment (sprankhub) [reply Test/Integration/AddScriptTest.php]: I asked to add the EU option. What I meant is: We should make sure that the script is inserted on the page no matter if the EU option is set or not. We don't need to test for the actual HTML as this is done by the `ScriptGeneratorTest`. But we should e.g. add `testScriptAddedOnHomepageWithEuCdn`, which tests if the script is inserted on the homepage when the EU CDN is enabled.

File: Test/Integration/AddScriptTest.php [hunk 10]
```diff
@@ -14,52 +14,64 @@ class AddScriptTest extends AbstractController
...
+        self::assertStringContainsString($this->script, $this->getResponse()->getBody());
+        self::assertEquals('<script id="Cookiebot" data-cfasync="false" src="https://consent.cookiebot.eu/uc.js" data-cbid=""  type="text/javascript" async></script>',
+            $this->script);
     }
 
     /**
-     * @magentoDataFixture   Magento/Catalog/_files/product_simple.php
-     * @magentoConfigFixture default/web/cookiebot/enabled 1
-     * @magentoConfigFixture default/web/cookiebot/id 123-456-789
+     * @magentoApiDataFixture   Magento/Catalog/_files/product_simple.php
```
Comment (sprankhub) [inline Test/Integration/AddScriptTest.php]:
```suggestion
     * @magentoDataFixture   Magento/Catalog/_files/product_simple.php
```

I guess this should be `magentoDataFixture`? Why with Api?

File: Test/Integration/Model/ScriptGeneratorTest.php (hunk 8 above)
Comment (sprankhub) [reply Test/Integration/Model/ScriptGeneratorTest.php]: Why two spaces? I don't see two spaces in the generator.

I would just test if `https://consent.cookiebot.com/uc.js` is included in the generated script. Nothing fancy, simply check if the right domain is included in the generated script.

**Review by sprankhub — CHANGES_REQUESTED** [review]

File: Test/Integration/Model/ConfigTest.php (hunk 9 above)
Comment (iranimij) [reply Test/Integration/Model/ConfigTest.php:57]: It's already been tested in scriptGeneratorTest and when those tests are true it means that the config is working fine. should we write duplicate tests? the thing that you want is part of the tests that are in ScriptGeneratorTest.php

File: Test/Integration/AddScriptTest.php (hunk 10 above)
Comment (iranimij) [reply Test/Integration/AddScriptTest.php]: because it just work with API, it got fail when it was without api

File: Test/Integration/Model/ConfigTest.php (hunk 9 above)
Comment (sprankhub) [reply Test/Integration/Model/ConfigTest.php:57]: This test is for a completely different class. So yes, I think it makes sense to have those tests. Plus, they are incredibly easy to add.

File: Test/Integration/AddScriptTest.php (hunk 7 above)
Comment (iranimij) [reply Test/Integration/AddScriptTest.php]: Actually I think you are talking about adding an extra duplicate test, since we also tested it in another file and I think it would extra duplicate test if we again here test the script tag in the html. what's you idea? 
the steps would be like
1_Test if script tag is in the html page
2_Test if eu cdn is true and works fine and Again if the test script tag is in the html page
3_In future for each single configuation we also want to check if script tag is in html page separately?

File: Test/Integration/AddScriptTest.php (hunk 10 above)
Comment (sprankhub) [reply Test/Integration/AddScriptTest.php]: Looks like it worked with `magentoDataFixture` before, and it still works with it for me. Also, reading through https://developer.adobe.com/commerce/webapi/get-started/web-api-functional-testing/#custom-annotations-for-data-fixtures, it doesn't sound like we should use `magentoApiDataFixture`.

File: Test/Integration/AddScriptTest.php (hunk 7 above)
Comment (sprankhub) [reply Test/Integration/AddScriptTest.php]: Theoretically, it could be that the script is included in the HTML with the flag disabled, but is not included in the HTML if the flag is enabled. So these tests do have their value. But I understand that you feel like it's overkill. I'll go ahead and revert the changes from this file then.

**Review by sprankhub — APPROVED** [review]
